Improve payroll modules

lomamech opened the pull request:
- Improve the way to send selected employee, send a array who contains the
  reference of selected Employees
- Deleting the request that runs in a loop,
- Implementation of function commitment who manage Data that will be used
  to execute the transaction to support the employees
- Update Integration Test for Payroll, Update sending data for Make
  Commitment
- Update payslip Generator API, Improve the request Perfomance
- Update Payslip Design of Payslip
- Improve the generated of report of Payroll
- Add two report of Payroll Report and Report of social Charge
- Add new function commitmentFunction for Generate data by Employees to
  make Commitment and Split the function Commitments
- Improved Payroll module, added the ability to display payroll bulletins
reports configured in multiple currencies into a single currency,
- Correction of conversion errors

Closes #2968

lomamech committed:
improvement(Payroll Modules)
- Improve the way to send selected employee, send a array who contains the
  reference of selected Employees
- Deleting the request that runs in a loop,
- Implementation of function commitment who manage Data that will be used
  to execute the transaction to support the employees

lomamech committed:
test(Payroll Modules)
- Update Integration Test for Payroll, Update sending data for Make
  Commitment

lomamech committed:
improvement(Payroll Modules)
- Update payslip Generator API, Improve the request Perfomance
- Update Payslip Design of Payslip

lomamech committed:
improvement(Payroll Report)
- Improve the generated of report of Payroll
- Add two report of Payroll Report and Report of social Charge

lomamech committed:
Update Writting Journal

lomamech committed:
improvement(Payroll Modules)
- Add new function commitmentFunction for Generate data by Employees to
  make Commitment and Split the function Commitments

lomamech committed:
improvement(Payroll Modules)
- Improved Payroll module, added the ability to display payroll bulletins
reports configured in multiple currencies into a single currency,
- Correction of conversion errors

lomamech committed:
test(Payroll Modules)
- Update the sending data for The Multi Payroll Configuration

lomamech committed:
improvement(Payroll)
- Remove not allowed spaces
- Remove exceeds maximum line length of 120
- Remove property declared in the upper scope
- Add Radix parameter
- Remove all unused parameter
- Use property shorthand

lomamech committed:
Fix the param already declared in the upper scope

lomamech committed:
test(Voucher Registry)
- Reset the value of numbers of Voucher Rows

lomamech committed:
test(Voucher)
- Reset Row Count in Vouchers

lomamech committed:
Sanitaze

jniles commented on client/src/modules/multiple_payroll/multiple_payroll.js:
```diff
@@ -256,6 +266,47 @@ function MultiplePayrollController(
+    // get All Employees Reference
+    const employeesRef = employees.map(emp => emp.reference);
+
+    if (employees.length) {
```
You probably don't need to check `employees.length` here.  `employees.forEach()` will skip an empty array.

lomamech replied:
@jniles it's just to warn the user that he has not selected any Employee
````
 if (employees.length) {
 ....
} else {
      Notify.danger('FORM.WARNINGS.NO_EMPLOYE_SELECTED');
}
````

jniles replied:
Ah, I see.  Maybe you should put that as the first thing checked? 

EDIT: as in:
```js
if (!employees.length) {
  Notify.warn(/*...*/);
}

// other code...

```

jniles commented on client/src/modules/multiple_payroll/multiple_payroll.js:
```diff
@@ -256,6 +266,47 @@ function MultiplePayrollController(
+      employees.forEach(employee => {
+        employeeStatusId = parseInt(employee.status_id, 10);
+
+        if (employeeStatusId === 1) {
```
It looks like you want to conditionally set `invalid` if some `employeeStatusId` is 1.  A clearer way to write this is to use [`Array.prototype.some()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/some).  It will return true if a condition is met for one element in the array.  Like this:

```js
// returns true if one employee has status_id of 1
const invalid = employees.some(employee => parseInt(employee.status_id, 10) === 1);
```

It would probably make it clearer to rename the function something that other programmers can read.  Like:
```js
const isAwaitingPayment =employee => parseInt(employee.status_id, 10) === 1;
const invalid = employees.some(employee => isAwaitingPayment(employee));
// alternatively: const invalid = employees.some(isAwaitingPayment);
```

jniles commented on client/src/modules/multiple_payroll/multiple_payroll.js:
```diff
@@ -256,6 +266,47 @@ function MultiplePayrollController(
+    const employees = vm.gridApi.selection.getSelectedRows();
+
+    // get All Employees Reference
+    const employeesRef = employees.map(emp => emp.reference);
```
Is it possible for employees to share references?  In most modules (cash, invoice, patient, etc), we assume references are shared between projects.  Do employees behave the same way?

lomamech replied:
No, employees are not related to the Project, only to the Enterprise

jniles replied:
Perfect 👍

jniles commented on server/controllers/finance/exchange.js:
```diff
@@ -143,3 +144,29 @@ exports.delete = function del(req, res, next) {
 };
+
+// This query returns the current exchange rate of all currencies
+function getCurrentExchangeRateByCurrency(date) {
```
This function is called `getCurrentExchangeRateByCurrency()`, but it doesn't take a currency.  It should be renamed to something that reflects what it actually does.

lomamech replied:
To test this function I had to add in my test database a third currency '€' Euro, and the function sent the exchange rate for all currencies in relation to the currency of the company

jniles commented on server/controllers/finance/exchange.js:
```diff
@@ -143,3 +144,29 @@ exports.delete = function del(req, res, next) {
+
+// This query returns the current exchange rate of all currencies
+function getCurrentExchangeRateByCurrency(date) {
+  const exchangeDate = date || new Date();
```
By the way, in the newest JavaScript, you can set default parameters for functions with a new syntax.  Instead of:
```js
function getCurrentExchangeRateByCurrency(date) {
  const exchangeDate = date || new Date();
  /// etc...
```
... you can do the same thing with this syntax....
```js
function getCurrentExchangeRateByCurrency(date = new Date()) {
  // now use `date` instead of `exchangeDate`
```

jniles commented on server/controllers/payroll/multiplePayroll/makeCommitment.js:
```diff
@@ -6,274 +6,103 @@
+        db.exec(sqlGetRubricPayroll, [referenceEmployees, payrollConfigurationId]),
+        db.exec(sqlGetRubricConfig, [payrollConfigurationId]),
+        db.exec(sqlGetAccountPayroll, [payrollConfigurationId]),
+        Exchange.getCurrentExchangeRateByCurrency(),
```
This function may return multiple exchange rates.  Why?  Shouldn't we use only one?

lomamech replied:
This function returns the last exchange rate configured per currency, if the system outside the company currency use several other currencies, this function would return the exchange rate for each of the currencie

jniles commented on server/controllers/payroll/multiplePayroll/makeCommitment.js:
```diff
@@ -6,274 +6,103 @@
+
+      return queries;
+    })
+    .spread((rubricsEmployees, rubricsConfig, account, exchangeRates) => {
```
Cool use of `.spread()` 👍

jniles commented on server/controllers/payroll/multiplePayroll/setConfiguration.js:
```diff
@@ -94,17 +94,23 @@ function config(req, res, next) {
      * Recalculation of base salary on the basis of any holiday or vacation period,
      * where the percentages are respectively equal to 100% of the basic salary will
-     * remain equal to that defined at the level of the grade table BB
+     * remain equal to that defined at the level of the Holiday table
```
💃

jniles commented on server/controllers/payroll/multiplePayroll/setMultiConfiguration.js:
```diff
@@ -7,16 +7,20 @@
 function config(req, res, next) {
-  const dataEmployees = req.body.data;
+  const dataEmployees = req.body.data.employees;
+  const currencyId = req.body.data.currencyId;
```
Just FYI, you can rewrite this as:
```js
const { employees, currencyId } = req.body.data;
```

You'd have to rename `dataEmployees` as `employees` however.

jniles commented on server/controllers/payroll/reports/payrollReportSocialCharge.handlebars:
```diff
@@ -0,0 +1,71 @@
+                <td style="text-align: left;"><strong> {{display_name}} </strong></td>
+                {{#each rubricsChargeEnterprise}}
+                  <td style="text-align: right;">
+                    <div>
```
Do you need these extra `<div>` and `<span>`s?

jniles commented on server/controllers/payroll/reports/payrollReportSocialCharge.handlebars:
```diff
@@ -0,0 +1,71 @@
+                  </td>
+                {{/each}}
+                <td>
+                  {{currency (multiply somChargeEnterprise ../payrollPeriod.exchangeRate) ../payrollPeriod.currency}}
```
Is there anyway to compute these values in JS instead of in the Handlebars template?

jniles commented on server/controllers/payroll/reports/payrollReportSocialCharge.handlebars:
```diff
@@ -0,0 +1,71 @@
+          <tbody>
+            {{#each dataEmployees}}
+              <tr class="text-capitalize">
+                <td style="text-align: left;"> {{add @index 1}} </td>
```
It doesn't look like `text-align:left` will actually do anything here.  Do you need to have it?

jniles replied:
Also, is this data sorted? I worry that the `@index` will change depending on the sort order of the data. Can we use the employee `reference` instead?

lomamech replied:
The employees are displayed by alphabetical order, the employees who appear on this report are those who have been selected, the index only allows the order of employees and it can also allow to quickly know the number of employees select

jniles commented on server/controllers/payroll/multiplePayroll/makeCommitment.js:
```diff
@@ -6,274 +6,103 @@
+    .then(dataEmployees => {
+      data.employees = dataEmployees;
+
+      const queries = q.all([
```
Instead of declaring `queries` and immediately returning it, it makes more sense to me to immediately `return q.all([])`.

jniles commented on server/controllers/payroll/reports/payslipGenerator.js:
```diff
@@ -1,56 +1,57 @@
+  } else if (options.currency && !options.socialCharge) {
+    TEMPLATE = templatePayrollReport;
+
+    options.footerRight = footerRight;
```
You repeat `options.footerRight` several times here.  Could you set a default set of options, then modify them if they need to be modified?  Like:
```js

const DEFAULT_OPTS = {
  orientation : 'landscape',
  footerRight : '[page] / [toPage]',
  filename : 'FORM.LABELS.PAYSLIP',
  // all other default opts...
};

function build(...) {
  _.extend(options, DEFAULT_OPTS);

  if (options.currency_id && !options.socialCharge) {
    options.orientation = 'portrait';
  }
}
```

jniles commented on server/controllers/payroll/reports/payslipGenerator.js:
```diff
@@ -64,153 +65,171 @@ function build(req, res, next) {
+      return Exchange.getExchangeRate(data.enterprise.id, options.currency, new Date(data.payrollPeriod.dateTo));
+    })
+    .then(exchange => {
+      data.payrollPeriod.exchangeRate = parseInt(options.currency, 10) === parseInt(data.enterprise.currency_id, 10) ?
```
Is the `parseInt()` here really necessary? Does the database give you strings for the id?

jniles commented on server/controllers/payroll/reports/payslipGenerator.js:
```diff
@@ -64,153 +65,171 @@ function build(req, res, next) {
+
+      dataEmployees.forEach(employee => {
+        data.exchangeRatesByCurrency.forEach(exchange => {
+          employee.net_salary_equiv = parseInt(exchange.currency_id, 10) === parseInt(employee.currency_id, 10) ?
```
Man ... if you are doing `parseInt() === parseInt()` every line, let's just do it once:
```js
const exchangeCurrencyId = parseInt(exchange.currency_id, 10);
const employeeCurrencyId = parseInt(employee.currency_id, 10);
const isSameCurrency = exchangeCurrencyId  === employeeCurrencyId ;

employee.net_salary_equiv = isSameCurrency ? employee.net_salary / exchange.rate : employee.net_salary;
```

Even better, we could set the value to 1 if the same currency:
```js
const exchangeCurrencyId = parseInt(exchange.currency_id, 10);
const employeeCurrencyId = parseInt(employee.currency_id, 10);
const isSameCurrency = exchangeCurrencyId  === employeeCurrencyId ;

// set the denominator to 1 if they are the same currency.
const base = isSameCurrency ? exchange.rate : 1;

employee.net_salary_equiv =  employee.net_salary / base;
employee.daily_salary_equiv = employee.daily_salary / base;
// etc...
```

jniles commented on server/controllers/payroll/reports/payslipGenerator.js:
```diff
@@ -64,153 +65,171 @@ function build(req, res, next) {
+          employee.somRubNonTaxable_equiv = parseInt(exchange.currency_id, 10) === parseInt(employee.currency_id, 10) ?
+            employee.somRubNonTaxable / exchange.rate : employee.somRubNonTaxable;
+          employee.somChargeEnterprise_equiv =
+            parseInt(exchange.currency_id, 10) === parseInt(employee.currency_id, 10) ?
```
You could do a similar thing to that `parseInt()` trick here.

jniles reviewed (commented):
@lomamech, this was a lot of changes, and I'm not sure I've gotten all of them.  Let me know when you are done with this review and I'll make an effort to take a look at them again.

lomamech committed:
improvement(Payroll)
- Use Array.prototype.some instead using forEach with alternative
  condition

lomamech committed:
improvement(Payroll)
- Set Default Parametres for set Date

lomamech committed:
improvement(Payroll Modules)
- Set the check of Employees in the top
- Improve the Codes
- Remove not necessary <div><span> in Handlebars

lomamech committed:
Sanitaze code not use parseInt for Enterprise Currency Id

jniles commented on test/end-to-end/vouchers/vouchers.spec.js:
```diff
@@ -6,12 +6,12 @@ const GridRow = require('../shared/GridRow');
   before(() => helpers.navigate('vouchers'));
 
-  it(`displays ${NUM_VOUCHERS} vouchers on the page`, () => {
+  it(`displays Above ${NUM_VOUCHERS} vouchers on the page`, () => {
```
💃

jniles commented on test/end-to-end/vouchers/vouchers.spec.js:
```diff
@@ -6,12 +6,12 @@ const GridRow = require('../shared/GridRow');
 
 describe('Voucher Registry', () => {
-  const NUM_VOUCHERS = 28;
+  const NUM_VOUCHERS = 23;
```
We should probably figure out a better way to test the number of records downloaded in the future.

jniles commented on server/controllers/payroll/reports/payslipGenerator.js:
```diff
@@ -1,216 +1,238 @@
+  let template;
+  _.extend(options, DEFAULT_OPTS);
+
+  if (options.currency && options.socialCharge) {
```
Much cleaner!

jniles reviewed (approved):
Code looks fine to me.  I've also tried it out and it seems to work alright.  Good work!

bors r+
